File: src/main/java/com/bna/gac/services/ClientService.java
package com.bna.gac.services;

import com.bna.gac.dto.ClientDTO;
import com.bna.gac.dto.StatusUpdateDTO;
import org.springframework.data.domain.Page;
import org.springframework.transaction.annotation.Transactional;

import java.util.List;

public interface ClientService {

    Page<ClientDTO> findAll(int page, int size, String search, String type, Boolean active);

    ClientDTO create(ClientDTO dto);

    ClientDTO update(Long id, ClientDTO dto);

    ClientDTO getById(Long id);

    List<ClientDTO> getAll();

    @Transactional
    ClientDTO updateStatus(Long id, StatusUpdateDTO dto);

    void delete(Long id);

    @Transactional
    ClientDTO deactivate(Long id);

    @Transactional
    ClientDTO reactivate(Long id);
}

File: frontend/src/app/clients/client.service.ts
import { Injectable } from "@angular/core";
import { HttpClient, HttpParams } from "@angular/common/http";
import { Observable } from "rxjs";
import { Client, buildClientPayload, TypeClient, PagedResponse } from "./client.model";
import { environment } from "../../environments/environment";
@Injectable({ providedIn: "root" })
export class ClientService {
  private readonly baseUrl = `${environment.apiUrl}/clients`;
  constructor(private http: HttpClient) {}
  getAll(page=0,size=10,search?: string,type?: TypeClient|"",active?: boolean): Observable<PagedResponse<Client>> {
    let params = new HttpParams().set("page",page.toString()).set("size",size.toString());
    if (search) params = params.set("search",search);
    if (type)   params = params.set("type",type);
    if (active !== undefined) params = params.set("active",active.toString());
    return this.http.get<PagedResponse<Client>>(this.baseUrl,{params});
  }
  getById(id: number): Observable<Client> { return this.http.get<Client>(`${this.baseUrl}/${id}`); }
  create(client: Partial<Client> & { typeClient: TypeClient }): Observable<Client> { return this.http.post<Client>(this.baseUrl, buildClientPayload(client)); }
  update(id: number, client: Partial<Client> & { typeClient: TypeClient }): Observable<Client> { return this.http.put<Client>(`${this.baseUrl}/${id}`, buildClientPayload(client)); }
  delete(id: number): Observable<void> { return this.http.delete<void>(`${this.baseUrl}/${id}`); }
  deactivate(id: number): Observable<Client> { return this.http.patch<Client>(`${this.baseUrl}/${id}/deactivate`,{}); }
  reactivate(id: number): Observable<Client> { return this.http.patch<Client>(`${this.baseUrl}/${id}/reactivate`,{}); }
}
